File: Year2023/src/Day7.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.function.Function;

/**
 * <a href="https://adventofcode.com/2023/day/7">Day 7</a>
 */
public class Day7 extends Day.IntDay {
    @Override
    public int run1Int(List<String> input) {
		return run(this::getCardValue, false);
    }

    @Override
    public int run2Int(List<String> input) {
		return run(card -> card == 'J' ? 1 : getCardValue(card), true);
    }

	private int run(Function<Character, Integer> cardValueGetter, boolean jokers) {
		List<Hand> hands = new ArrayList<>();
		for(String s : getInput()) {
			String[] parts = s.split(" ");
			int bid = Utils.fastParseInt(parts[1]);
			int firstCard = cardValueGetter.apply(parts[0].charAt(0));
			HandType type = getHandType(parts[0], jokers);
			hands.add(new Hand(parts[0], type, bid));
		}

		hands.sort((a, b) -> {
			if(a.type != b.type) {
				return a.type.compareTo(b.type);
			}
			for(int i = 0; i < a.cards.length(); i++) {
				int aVal = cardValueGetter.apply(a.cards.charAt(i));
				int bVal = cardValueGetter.apply(b.cards.charAt(i));
				if(aVal != bVal) {
					return bVal - aVal;
				}
			}
			return 0;
		});

		Collections.reverse(hands);
		int sum = 0;
		for(int i = 0; i < hands.size(); i++) {
			sum += hands.get(i).bid *(i + 1);
		}
		return sum;
	}

	private int getCardValue(char card) {
		return switch(card) {
			case 'A' -> 14;
			case 'K' -> 13;
			case 'Q' -> 12;
			case 'J' -> 11;
			case 'T' -> 10;
			default -> Utils.fastParseInt(card);
		};
	}

	private HandType getHandType(String hand, boolean jokers) {
		if(hand.contains("J") && jokers) {
			hand = replaceJoker(hand);
		}

		return switch(countUniqueCards(hand)) {
			case 1 -> HandType.FIVE_OF_A_KIND;
			case 2 -> evaluateTwoUnique(hand);
			case 3 -> evaluateThreeUnique(hand);
			case 4 -> HandType.ONE_PAIR;
			default -> HandType.HIGH_CARD;
		};
	}

	private String replaceJoker(String hand) {
		Map<Character, Long> frequencyMap = countCardFrequency(hand.replace("J", ""));
		char mostFrequentCard = getMostFrequentCard(frequencyMap);
		return hand.replace("J", String.valueOf(mostFrequentCard));
	}

	private int countUniqueCards(String hand) {
		int numUnique = 0;
		boolean[] visited = new boolean[128]; // Assuming ASCII characters
		for(char character : hand.toCharArray()) {
			if(!visited[character]) {
				visited[character] = true;
				numUnique++;
			}
		}
		return numUnique;
	}

	private Map<Character, Long> countCardFrequency(String hand) {
		Map<Character, Long> map = new HashMap<>();
		for(char c : hand.toCharArray()) {
			map.merge(c, 1L, Long::sum);
		}
		return map;
	}

	private char getMostFrequentCard(Map<Character, Long> frequencyMap) {
		char mostFrequentCard = 0;
		long maxFrequency = 0;
		for(Entry<Character, Long> entry : frequencyMap.entrySet()) {
			if(entry.getValue() > maxFrequency) {
				mostFrequentCard = entry.getKey();
				maxFrequency = entry.getValue();
			}
		}
		return mostFrequentCard;
	}

	private HandType evaluateTwoUnique(String hand) {
		int numFirst = 0;
		for(char c : hand.toCharArray()) {
			if(c == hand.charAt(0)) {
				numFirst++;
			}
		}
		return(numFirst == 4 || numFirst == 1) ?
			HandType.FOUR_OF_A_KIND : HandType.FULL_HOUSE;
	}

	private HandType evaluateThreeUnique(String hand) {
		int[] freq = new int[128];
		for(char character : hand.toCharArray()) {
			freq[character]++;
		}

		int countOfThree = 0;
		int countOfTwo = 0;
		int countOfOne = 0;

		for(int i = 0; i < 128; i++) {
			if(freq[i] == 3) countOfThree++;
			else if(freq[i] == 2) countOfTwo++;
			else if(freq[i] == 1) countOfOne++;
		}

		if(countOfThree == 1 && (countOfTwo == 1 || countOfOne == 2)) {
			return HandType.THREE_OF_A_KIND;
		} else if(countOfThree == 1 && countOfOne == 1) {
			return HandType.FULL_HOUSE;
		} else {
			return HandType.TWO_PAIR;
		}
	}

	record Hand(String cards, HandType type, int bid) {}

	enum HandType {
		FIVE_OF_A_KIND,
		FOUR_OF_A_KIND,
		FULL_HOUSE,
		THREE_OF_A_KIND,
		TWO_PAIR,
		ONE_PAIR,
		HIGH_CARD
	}
}

File: src/2023/kotlin/day7.kt
import dev.rdh.aoc.*
import java.util.Arrays

private fun PuzzleInput.part1(): Any? {
    return run(jokers = false)
}

private fun PuzzleInput.part2(): Any? {
    return run(jokers = true)
}

private fun PuzzleInput.run(jokers: Boolean): Int {
    return lines.map {
        val parts = it.split(" ")
        Hand(parts[0], getHandType(parts[0], jokers), parts[1].toInt())
    }.sortedWith { a, b ->
        if (a.type != b.type) {
            return@sortedWith a.type.compareTo(b.type)
        }

        val (a, b) = listOf(a, b).map { it.cards.map { getCardValue(it, jokers) }.toIntArray() }
        return@sortedWith Arrays.mismatch(a, b).let {
            b[it] - a[it]
        }
    }.reversed().mapIndexed { i, h ->
        h.bid * (i + 1)
    }.sum()
}

private fun getCardValue(card: Char, jokers: Boolean): Int {
    return when (card) {
        'A' -> 14
        'K' -> 13
        'Q' -> 12
        'J' -> if (jokers) 1 else 11
        'T' -> 10
        else -> card.digitToInt()
    }
}

private fun getHandType(hand: String, jokers: Boolean): HandType {
    var hand = hand
    if (hand.contains("J") && jokers) {
        hand = replaceJoker(hand)
    }

    return when (hand.toCharArray().distinct().size) {
        1 -> HandType.FIVE_OF_A_KIND
        2 -> evaluateTwoUnique(hand)
        3 -> evaluateThreeUnique(hand)
        4 -> HandType.ONE_PAIR
        else -> HandType.HIGH_CARD
    }
}

private fun replaceJoker(hand: String): String {
    val frequencyMap = hand.replace("J", "").groupingBy { it }.eachCount().toMutableMap()
    val mostFrequentCard = frequencyMap.maxByOrNull { it.value }?.key
    return hand.replace('J', mostFrequentCard ?: 'A') // if all jokers, replace with A
}

private fun evaluateTwoUnique(hand: String): HandType {
    return hand.count { it == hand[0] }.let {
        if (it == 4 || it == 1) HandType.FOUR_OF_A_KIND else HandType.FULL_HOUSE
    }
}

private fun evaluateThreeUnique(hand: String): HandType {
    val freq = IntArray(128) {
        hand.count { c -> c.code == it }
    }

    val (countOfOne, countOfTwo, countOfThree) = listOf(1, 2, 3).map { i ->
        freq.count { i == it }
    }

    return when {
        countOfThree == 1 && (countOfTwo == 1 || countOfOne == 2) -> HandType.THREE_OF_A_KIND
        countOfThree == 1 && countOfOne == 1 -> HandType.FULL_HOUSE
        else -> HandType.TWO_PAIR
    }
}

private data class Hand(val cards: String, val type: HandType, val bid: Int)

private enum class HandType {
    FIVE_OF_A_KIND,
    FOUR_OF_A_KIND,
    FULL_HOUSE,
    THREE_OF_A_KIND,
    TWO_PAIR,
    ONE_PAIR,
    HIGH_CARD
}

fun main() {
    val input = PuzzleInput(2023, 7)

    var start = System.nanoTime()
    var result = input.part1()
    var end = System.nanoTime()
    println("--- Part 1: %.2fms ---".format((end - start) / 1e6))
    println(result)

    start = System.nanoTime()
    result = input.part2()
    end = System.nanoTime()
    println("--- Part 2: %.2fms ---".format((end - start) / 1e6))
    println(result)
    println("----------------------")
}